src/synthesized_insight/metrics/statistical_tests.py

```diff
@@ -0,0 +1,317 @@
+        corr, p_value = spearmanr(x, y, nan_policy='omit')
+        if p_value is not None and p_value <= self.max_p_value:
+            return corr, p_value
+        else:
+            return None, p_value
+
+
+class BootstrapTest(TwoColumnMetricTest):
+    """Performs bootstrap test
+
+    Args:
+        metric_cls_obj: The instantiated metric object for which the test if performed
+        binned: If binning is done
+    """
+    name = "bootstrap_test"
+
+    def __init__(self,
+                 metric_cls_obj: TwoColumnMetric,
+                 check: ColumnCheck = None,
+                 alternative: str = 'two-sided',
+                 binned: bool = False):
+        super().__init__(check, alternative)
+        self.metric_cls_obj = metric_cls_obj
+        self.binned = binned
+
+    @classmethod
+    def check_column_types(cls, check: ColumnCheck, sr_a: pd.Series, sr_b: pd.Series) -> bool:
+        return True  # metric object provided during initialization will perform the check while computing the metrics
+
+    def _bootstrap_pvalue(self,
+                          t_obs: float,
+                          t_distribution: pd.Series) -> float:
+        """
+        Calculate a p-value using a bootstrapped test statistic distribution
+
+        Args:
+            t_obs: Observed value of the test statistic.
+            t_distribution: Samples of test statistic distribution under the null hypothesis.
+
+        Returns:
+            The p-value under the null hypothesis.
+        """
+
+        n_samples = len(t_distribution)
+
+        if self.alternative == 'two-sided':
+            p = np.sum(np.abs(t_distribution) >= np.abs(t_obs)) / n_samples
+
+        elif self.alternative == 'greater':
+            p = np.sum(t_distribution >= t_obs) / n_samples
+
+        else:
+            p = np.sum(t_distribution < t_obs) / n_samples
+
+        return p
+
+    def _compute_test(self, sr_a: pd.Series, sr_b: pd.Series) -> Tuple[Union[int, float, None], Union[int, float, None]]:
+        """Return metric_value and p-value for this metric using a bootstrap test.
+
+        The null hypothesis is that both data samples are from the same distribution."""
+
+        metric_value = self.metric_cls_obj(sr_a, sr_b)
+        if metric_value is None:
+            return None, None
+
+        if not self.binned:
+            ts_distribution = bootstrap_statistic((sr_a, sr_b),
+                                                  lambda x, y: self.metric_cls_obj(x, y),
+                                                  n_samples=1000)
+        else:
+            ts_distribution = bootstrap_binned_statistic((sr_a, sr_b),
+                                                         lambda x, y: self.metric_cls_obj(x, y),
+                                                         n_samples=1000)
+        p_value = self._bootstrap_pvalue(metric_value, ts_distribution)
+        return metric_value, p_value
+
+
+class PermutationTest(TwoColumnMetricTest):
```

Comment:
shall we try and remove this?
```
class PermutationTest(_Metric):
```
and then write our own `__call__` and remove/replace `_check_column_types`

Reply:
same with bootstrap

Reply:
Modified as suggested